libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1755 @@
+            }
+            
+            token = self.next(); // check if next is right paren or comma
+            if token.token_type != TokenType::COMMA {
+                // if the next token is not comma, must be right paren
+                break;
+            } 
+            token = self.next(); // if its a comma, then there are more args to parse
+        }
+        
+
+        if token.token_type != TokenType::RIGHTPAREN {
+            return Err("Function arguments must be follow by a right paren")
+        }
+
+        token = self.next(); // move to arrow
+
+        if token.token_type != TokenType::ARROW {
+            return Err("Function must have an arrow before return monotype")
+        }
+
+        // recursively parse the function's return type
+        let return_type = self.parse_monotype(); 
+
+        if let Ok(return_val) = return_type {
+            Ok(MonoType::Fun(Box::new(Function{
+                req: req_args,
+                opt: opt_args,
+                pipe: pipe_arg,
+                retn: return_val,
+            }))) 
+        } else {
+            return Err("Function must have a valid return type")
+        }
+    }
+
+    // parse_required_optional parses a single required or optional argument for the function monotype
+    fn parse_required_optional(& mut self, token: &Token) -> Result<(String, MonoType), &'static str> {
+        let mut arg_var = String::new(); 
+
+        match &token.text {
+            None => {
+                return Err("Invalid format for required arguments")
+            },
+            Some(var) => {
+                arg_var = var.to_string();
+            }
+        }
+
+        let token = self.next(); 
+        if token.token_type != TokenType::COLON {
+            return Err("Invalid format for required arguments")
+        }
+
+        let monotype = self.parse_monotype(); 
+
+        match monotype {
+            Err(e) => Err(e), 
+            Ok(monotype) => Ok((arg_var, monotype))
+        }
+    }
+
+    // parse_pipe parses a single pipe argument for the function monotype
+    fn parse_pipe(& mut self) -> Option<Property> {
+        let mut token = self.peek(); 
+
+        let mut string = None;  
+        let mut monotype = Err("No monotype found"); 
+        if token.token_type == TokenType::IDENTIFIER {
+            token = self.next(); 
+
+            string = token.text; 
+        }
+
+        let next_token = self.peek();
+        if next_token.token_type == TokenType::COLON {
+            self.next(); 
+            monotype = self.parse_monotype(); 
+        }
+
+        match monotype {
+            // if there's no monotype, there's no pipe argument 
+            Err(_e) => None, 
+            Ok(monotype) => {
+                // string is optional, so still return Property
+                match string {
+                    None => Some(Property {
+                        k: String::new(),
```

Comment:
I think this should be the actual string `"<-"` since property names aren't optional.

Reply:
Do you only want the pipe to be included in the string in cases where the pipe argument doesn't have a variable name?